Bio/Seq.py

```diff
@@ -1000,9 +1014,25 @@ def translate(self, table="Standard", stop_symbol="*", to_stop=False,
                 # The same table can be used for RNA or DNA (we use this for
                 # translating strings).
                 codon_table = CodonTable.ambiguous_generic_by_id[table_id]
-        protein = _translate_str(str(self), codon_table,
-                                 stop_symbol, to_stop, cds)
-        if stop_symbol in protein:
+
+        # Deal with gaps for translation
+        if hasattr(self.alphabet, "gap_char"):
+            if not gap:
+                gap = self.alphabet.gap_char
+            elif gap != self.alphabet.gap_char:
+                raise ValueError("Gap {0!r} does not match {1!r} from alphabet".format(
+                    gap, self.alphabet.gap_char))
+
+        protein = _translate_str(str(self), codon_table, stop_symbol, to_stop,
+                                 cds, gap=gap)
+        if gap and \
+                gap in protein and \
+                stop_symbol in protein:
+            alphabet = Alphabet.HasStopCodon(Alphabet.Gapped(codon_table.protein_alphabet,
+                                                             gap_char=gap))
```

Comment:
You're not using the ``stop_symbol`` in ``HasStopCodon``, so that would default to ``*``.

You've also split the if statement up into more lines than needed with ``\`` characters.

I'd probably do this as two separate if statements,

```
if gap and gap in protein:
    alphabet = Alphabet.Gapped(codon_table.protein_alphabet, gap)
else:
    alphabet = codon_table.protein_alphabet

if stop_symbol in protein:
    alphabet = Alphabet.HasStopCodon(alphabet, stop_symbol)
```

Or the other way round depending which order seems best for the alphabet class wrapping horribleness?